File: android-beans/src/test/java/rocks/frieler/android/beans/scopes/activity/ForegroundActivityHolderTest.java
package rocks.frieler.android.beans.scopes.activity;

import android.app.Activity;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.CoreMatchers.sameInstance;
import static org.junit.Assert.assertThat;

@RunWith(MockitoJUnitRunner.class)
public class ForegroundActivityHolderTest {
    private ForegroundActivityHolder foregroundActivityHolder = new ForegroundActivityHolder();

    @Mock
    private Activity activity, anotherActivity;

    @Test
    public void testForegroundActivityHolderHoldsCreatedActivity() {
        foregroundActivityHolder.onActivityCreated(activity, null);

        assertThat(foregroundActivityHolder.getCurrentActivity(), is(sameInstance(activity)));
    }

    @Test
    public void testForegroundActivityHolderHoldsStartedActivity() {
        foregroundActivityHolder.onActivityStarted(activity);

        assertThat(foregroundActivityHolder.getCurrentActivity(), is(sameInstance(activity)));
    }

    @Test
    public void testForegroundActivityHolderHoldsResumedActivity() {
        foregroundActivityHolder.onActivityResumed(activity);

        assertThat(foregroundActivityHolder.getCurrentActivity(), is(sameInstance(activity)));
    }

    @Test
    public void testForegroundActivityHolderResetsActivityWhenPaused() {
        foregroundActivityHolder.onActivityResumed(activity);

        foregroundActivityHolder.onActivityPaused(activity);

        assertThat(foregroundActivityHolder.getCurrentActivity(), is(nullValue()));
    }

    @Test
    public void testForegroundActivityHolderKeepsActivityWhenAnotherActivityIsPaused() {
        foregroundActivityHolder.onActivityResumed(activity);

        foregroundActivityHolder.onActivityPaused(anotherActivity);

        assertThat(foregroundActivityHolder.getCurrentActivity(), is(sameInstance(activity)));
    }

    @Test
    public void testForegroundActivityHolderResetsActivityWhenStopped() {
        foregroundActivityHolder.onActivityResumed(activity);

        foregroundActivityHolder.onActivityStopped(activity);

        assertThat(foregroundActivityHolder.getCurrentActivity(), is(nullValue()));
    }

    @Test
    public void testForegroundActivityHolderKeepsActivityWhenAnotherActivityIsStopped() {
        foregroundActivityHolder.onActivityResumed(activity);

        foregroundActivityHolder.onActivityStopped(anotherActivity);

        assertThat(foregroundActivityHolder.getCurrentActivity(), is(sameInstance(activity)));
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/scopes/activity/ForegroundActivityHolderTest.kt
package rocks.frieler.android.beans.scopes.activity

import android.app.Activity
import assertk.assertThat
import assertk.assertions.isNull
import assertk.assertions.isSameAs
import assertk.assertions.isSameInstanceAs
import org.mockito.kotlin.mock
import org.junit.jupiter.api.Test

class ForegroundActivityHolderTest {
	private val foregroundActivityHolder = ForegroundActivityHolder()

	private val activity: Activity = mock()

	private val anotherActivity: Activity = mock()

	@Test
	fun `ForegroundActivityHolder holds created Activity`() {
		foregroundActivityHolder.onActivityCreated(activity, null)

		assertThat(foregroundActivityHolder.currentActivity).isSameInstanceAs(activity)
	}

	@Test
	fun `ForegroundActivityHolder holds started Activity`() {
		foregroundActivityHolder.onActivityStarted(activity)

		assertThat(foregroundActivityHolder.currentActivity).isSameInstanceAs(activity)
	}

	@Test
	fun `ForegroundActivityHolder holds resumed Activity`() {
		foregroundActivityHolder.onActivityResumed(activity)

		assertThat(foregroundActivityHolder.currentActivity).isSameInstanceAs(activity)
	}

	@Test
	fun `ForegroundActivityHolder resets Activity when paused`() {
		foregroundActivityHolder.onActivityResumed(activity)
		foregroundActivityHolder.onActivityPaused(activity)

		assertThat(foregroundActivityHolder.currentActivity).isNull()
	}

	@Test
	fun `ForegroundActivityHolder keeps Activity when another Activity is paused`() {
		foregroundActivityHolder.onActivityResumed(activity)
		foregroundActivityHolder.onActivityPaused(anotherActivity)

		assertThat(foregroundActivityHolder.currentActivity).isSameInstanceAs(activity)
	}

	@Test
	fun `ForegroundActivityHolder resets Activity when stopped`() {
		foregroundActivityHolder.onActivityResumed(activity)
		foregroundActivityHolder.onActivityStopped(activity)

		assertThat(foregroundActivityHolder.currentActivity).isNull()
	}

	@Test
	fun `ForegroundActivityHolder keeps Activity when another Activity is stopped`() {
		foregroundActivityHolder.onActivityResumed(activity)
		foregroundActivityHolder.onActivityStopped(anotherActivity)

		assertThat(foregroundActivityHolder.currentActivity).isSameInstanceAs(activity)
	}
}
